### PR TITLE
Adding database specific settings, cancel migration and other dark UI fixes

### PR DESCRIPTION
**Pull request by aasimkhan30**
Now we can specify the target database name and database specific file path. I have added the same for blob containers and file shares but that workflow is disabled for now.

image

Summary page info:
image

Dashboard  Dark UI fixes. It is still WIP. 
image
Dark UI
image

Fixed Target Azure SQL Instance name  (Previously it was showing database name)
image

Added Cancel Migration Button: 
image

### CODE REVIEW

File: extensions/sql-migration/src/api/azure.ts [hunk 1]
```diff
@@ -220,6 +220,17 @@ export async function startMigrationCutover(account: azdata.Account, subscriptio
 	return response.response.data.value;
 }
 
+export async function stopMigration(account: azdata.Account, subscription: Subscription, migrationStatus: DatabaseMigration): Promise<any> {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:223]: Don't use any please

File: extensions/sql-migration/src/constants/strings.ts [hunk 2]
```diff
@@ -87,7 +87,21 @@ export const INVALID_FILESHARE_ERROR = localize('sql.migration.invalid.fileShare
...
+export function TARGET_NETWORK_SHARE_LOCATION(dbName: string): string {
+	return localize('sql.migration.network.share.location', "Network share location to read backups for database ‘{0}’", dbName);
+}
+export function TARGET_FILE_SHARE(dbName: string): string {
+	return localize('sql.migration.file.share', "Select the file share that contains the backup files for ‘{0}’", dbName);
+}
+export function TARGET_BLOB_CONTAINER(dbName: string): string {
+	return localize('sql.migration.blob.container', "Select the container that contains the backup files for ‘{0}’", dbName);
+}
+export const ENTER_NETWORK_SHARE_INFORMATION = localize('sql.migration.enter.network.share.information', "Enter network share path information for selected database(s)");
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/constants/strings.ts:102]: Avoid using `(s)` - it's difficult to translate properly. Generally it's fine just to pluralize it.

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts [hunk 3]
```diff
@@ -290,6 +291,27 @@ export class MigrationCutoverDialog {
 			}
 		});
 
+		this._cancel = this._view.modelBuilder.button().withProps({
+			iconPath: IconPathHelper.discard,
+			iconHeight: '16px',
+			iconWidth: '16px',
+			label: 'Cancel Migration',
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts:298]: localize

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 4]
```diff
@@ -102,7 +104,7 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 	private _skuRecommendations: SKURecommendations | undefined;
 	private _assessmentResults: mssql.SqlMigrationAssessmentResultItem[] | undefined;
 
-
+	public _refreshDatabaseBackupPage!: boolean;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:107]: Public vars shouldn't start with `_`

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 5]
```diff
@@ -105,6 +108,8 @@ export class DatabaseBackupPage extends MigrationWizardPage {
 			}
 		});
 
+		networkShareButton.checked = true;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:111]: Just put this in the initial withProps

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 6]
```diff
@@ -465,8 +516,10 @@ export class DatabaseBackupPage extends MigrationWizardPage {
 					if ((<azdata.CategoryValue>this._blobContainerStorageAccountDropdown.value).displayName === constants.NO_STORAGE_ACCOUNT_FOUND) {
 						errors.push(constants.INVALID_STORAGE_ACCOUNT_ERROR);
 					}
-					if ((<azdata.CategoryValue>this._blobContainerBlobDropdown.value).displayName === constants.NO_BLOBCONTAINERS_FOUND) {
-						errors.push(constants.INVALID_BLOBCONTAINER_ERROR);
+					for (let i = 0; i < this._blobContainerDropdowns.length; i++) {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:519]: Why do you need to iterate over multiple? Shouldn't this be the only one?

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 7]
```diff
@@ -476,8 +529,10 @@ export class DatabaseBackupPage extends MigrationWizardPage {
 					if ((<azdata.CategoryValue>this._fileShareStorageAccountDropdown.value).displayName === constants.NO_STORAGE_ACCOUNT_FOUND) {
 						errors.push(constants.INVALID_STORAGE_ACCOUNT_ERROR);
 					}
-					if ((<azdata.CategoryValue>this._fileShareFileShareDropdown.value).displayName === constants.NO_FILESHARES_FOUND) {
-						errors.push(constants.INVALID_FILESHARE_ERROR);
+					for (let i = 0; i < this._fileShareDropdowns.length; i++) {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:532]: Same here - when would the message not be the only value?

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 8]
```diff
@@ -508,28 +563,56 @@ export class DatabaseBackupPage extends MigrationWizardPage {
...
 		this._blobContainerSubscriptionDropdown.validate();
 		this._blobContainerStorageAccountDropdown.validate();
-		this._blobContainerBlobDropdown.validate();
+		this._blobContainerDropdowns.forEach((dropdown) => {
+			dropdown.validate();
+		});
 		this._fileShareSubscriptionDropdown.validate();
 		this._fileShareStorageAccountDropdown.validate();
-		this._fileShareFileShareDropdown.validate();
+		this._fileShareDropdowns.forEach((dropdown) => {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:591]: FYI you don't need to add the parens when you have just a single parameter without its type

File: extensions/sql-migration/src/wizard/migrationModePage.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import { MigrationWizardPage } from '../models/migrationWizardPage';
+import { MigrationCutover, MigrationStateModel, StateChangeEvent } from '../models/stateMachine';
+import * as constants from '../constants/strings';
+
+export class MigrationModePage extends MigrationWizardPage {
+	constructor(wizard: azdata.window.Wizard, migrationStateModel: MigrationStateModel) {
+		super(wizard, azdata.window.createWizardPage(constants.DATABASE_BACKUP_MIGRATION_MODE_LABEL), migrationStateModel);
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/migrationModePage.ts:14]: Would suggest giving the page a (non-localized) name for telemetry purposes

File: extensions/sql-migration/src/wizard/migrationModePage.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+		const onlineButton = view.modelBuilder.radioButton().withProps({
+			label: constants.DATABASE_BACKUP_MIGRATION_MODE_ONLINE_LABEL,
+			name: buttonGroup,
+			checked: true
+		}).component();
+
+		this.migrationStateModel._databaseBackup.migrationCutover = MigrationCutover.ONLINE;
+
+		onlineButton.onDidChangeCheckedState((e) => {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/migrationModePage.ts:46]: All of these event handlers should be added to a disposable store and disposed of properly when the wizard is closed (or whenever they're not needed anymore)

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-migration/src/wizard/migrationModePage.ts (hunk 10 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/wizard/migrationModePage.ts:46]: Will do this in a future PR.

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts (hunk 6 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/wizard/databaseBackupPage.ts:519]: I will break the loop after first invalid dropdown in the array.

**Review by kburtram — APPROVED** [review]
LGTM

**Comment by kburtram** [comment]
@aasimkhan30 some of the links looks too dark for on the black background in dark theme.  are we getting those link colors from the theme settings?

**Comment by aasimkhan30** [comment]
@kburtram , the dark links are from the disabled buttons. 
Start cutover and cancel migration buttons are disabled when the migration is not in progress. Let me know if you are referring to some other links.

**Pull request merged**